### DISCUSSION TITLE
Copybook configuration without zowe profile

### DISCUSSION THREAD

**Discussion by abhimanDas** [IBM/zopeneditor-about#218] [Q&A]
Is there is any way we could fetch datasets from mainframe without creating a zowe profile. I tried configuring settings.json, but it throws an error everytime stating that a mvs type requires a valid zowe profile.

**Comment by FALLAI-Denis** [comment]
Hi @abhimanDas 

You should explain what prevents you from using a Zowe profile to access the z/OS mainframe and therefore the copybooks.
I guess it's the lack of z/OSMF services (z/OS communication system) or RSE services (historical IBM Eclipse platforms communication system).

Maybe you can try to create a Zowe zFTP type profile. This works at the Zowe Explorer level to access z/OS resources, instead of a zOSMF or RSE type profile, (you also need to install "Zowe Explorer Extension for FTP").
But I haven't tested the operation with Z Open Editor (which uses Zowe Explorer services to access z/OS resources).
A Zowe zFTP profile will allow you at least to download your copybooks locally with Zowe Explorer, then to tell Z Open Editor where to find them (via property-groups).

**Reply by abhimanDas** [reply]
Hi Dennis,

We are not using zowe layer to communicate with Z/OS as we have our own CLI to perform the communication. 
While using the IBM open Z Editor to open the cobol files, I figured we can use MVS or local types to configure the copybooks.
As I could not use zowe profiles, I was able to download the copybooks locally using our own cli and configure them up to use with Open Z Editor. 
**But I was wondering if there is any other way we could configure mvs type instead of using local, without the zowe dependency.**

**Comment by FALLAI-Denis** [comment]
@abhimanDas 

For information, I have already made a similar request:

[All languages - Remove dependency on Zowe Explorer](https://github.com/IBM/zopeneditor-about/issues/179)

**Comment by phaumer** [comment]
We have not yet added FTP as a protocol for resolving remote copybooks, but I think this would be an easy thing for us to add. 

When you mentioned "Zowe layer", note, that Zowe CLI profiles do not require any Zowe server components to be installed on z/OS. You need instead either RSE (available for free with IBM z/OS Explorer; a protocol that we support for all our IBM editors) or z/OSMF which is part of z/OS itself. 

We currently have no plans for moving away from Zowe CLI as it is a convenient open source API in an open source project that we are contributing to.  Zowe CLI and Zowe Explorer are extensible so you can implement your own Zowe CLI plugin that could add another protocol to the currently available ones (z/OSMF, FTP, SSH). If you donate such a plugin to Zowe we could discuss supporting it for Z Open Editor as well.

**Reply by FALLAI-Denis** [reply]
Zowe CLI FTP like Zowe CLI SSH do not correctly manage encoding (ccsid).

Again, Z Open Editor should be modular and not limit or impose services implementation choices that are not directly related to source code editing.

Whether z/OSMF or RSE is free of charge, is not a sufficient reason to impose them. There are other constraints to take into account such as not multiplying the solutions meeting the same need because this increases the administration costs. If a site has chosen a communication solution with the z/OS system, it should be free to continue using it with Z Open Editor.

Whenever Z Open Editor uses a service that is not directly related to source code editing, it should go through an abstraction API, which makes it possible to offer a turnkey solution by interfacing by default other extensions, but also to give the possibility to interface an extension of one's choice, or even to develop into a proprietary extension to implement the service.

We are very careful not to lock ourselves into an "all-in-one" solution, provided by a single supplier without alternatives.
We want the freedom to assemble our solution from bricks that we choose, not bricks that are imposed on us.
These bricks can also be under paid license, it is not a problem.
Moreover, for information, we have "Wazi" licenses in our IBM contract.

**Comment by phaumer** [comment]
But that abstraction layers and API exists and is documented here: https://github.com/zowe/vscode-extension-for-zowe/blob/master/docs/README-Extending.md#creating-an-extension-that-adds-a-data-provider

If more implementations of these interfaces appear I promise we will look at them as long as we are able try and test with them or a business partner relationship that will support it.
